File: matlab/Figure10_2to4.m
% Author: Kenji Kashima
% Date  : 2025/09/10
% Note  : Optimization Toolbox is required

clear;close all; rng(3); % random seed

% parameter setting
k_bar = 50;
k_bar_i = 1000;
sigma_param = 1.0; 

% transition probability matrix (P0)
P = [1/3,    1/3,     0,     0; 
      0  ,    1/3,   1/3,     0;
      0  ,    1/3,   1/3,   1/3;
      2/3,      0,   1/3,   2/3];

p_stationary = stationary_distribution(P);
disp('p_stationary for P^0:');
disp(p_stationary);

% accumulated transition probability
[m,n] =size(P);
P_accum = cumsum(P,1);

p_list = zeros(1,k_bar); 
p_list(1)=4; % start at 4
for i=2:k_bar
    u = rand;
    T = P_accum(:,p_list(i-1)); %transition probability
    for j = 1:m
        if u <= T(j)
             p_list(i) = j;
             break
        end
    end
end

figure('Name','Figure10.2(b)'); hold on; grid on;
stairs(p_list);
yticks(1:4);
xlim([0,k_bar]);
ylim([0.8,4.2]);
xlabel('$k$','Fontsize',16,'Interpreter','latex')
title('$P^0$','Fontsize',16,'Interpreter','latex')

%% inverse reinforcement learning

beta = 0.8;
invbeta = 1/beta;
cost_val = [1, 2, 3, 4] * sigma_param; 

% solve 10.23 L_KL
options = optimoptions('fminunc', 'Algorithm','quasi-newton',...
    'OptimalityTolerance',1e-10, 'FunctionTolerance',1e-14,...
    'MaxIterations',1000, 'MaxFunctionEvaluations',5000);
v_init_kl = ones(4,1)*5; % Initialization for minimization
V_star_optimal = fminunc(@(v_star)L_KL(v_star, beta, cost_val, P), v_init_kl, options); % minimization of L_KL

disp("Value function：");
disp(V_star_optimal);
z_opt = exp(-beta * V_star_optimal); % Calculate Z* from V*

% compute P_opt from Z*
P_opt = zeros(size(P));
for r_idx = 1:m 
    for c_idx = 1:n 
        P_opt(r_idx,c_idx) = P(r_idx,c_idx) * z_opt(r_idx) / (P(:,c_idx)' * z_opt); 
    end
end

fprintf("sigma: %.1f\nPpi:\n", sigma_param);
disp(P_opt);

p_opt_stationary = stationary_distribution(P);
disp("p_stationary for P_opt：");
disp(p_opt_stationary);


% accumulate P
P_accum = cumsum(P_opt, 1);

% simulation
p_list = zeros(k_bar_i+1, 1, 'int8');
p_list(1) = 4; % initial state
a_counts = [0; 0; 0; 1]; 
b_counts = [0; 0; 0; 0]; 
inv_l_hist = zeros(4, k_bar);

% constrains (you can change lb, to move the baseline)
options_v_fmincon = optimoptions('fmincon','Display','off'); 
lb_v = [12; 12; 12; 12]; % lower bound constrains

for k_i = 1:k_bar_i 
    % maximize the likelyhood L_IRL 
    v_irl_init = [1; 2; 3; 4]; % Initialization for minimization
    V_star_estimated_irl = fmincon(@(v_star_irl)L_IRL(v_star_irl, beta, a_counts, b_counts, P), v_irl_init,... % mimization of L_IRL
        [],[],[],[],lb_v,[],[],options_v_fmincon);
    
    inv_v_opt_irl = V_star_estimated_irl; 
    inv_z_opt_irl = exp(-beta * inv_v_opt_irl); 
    inv_l_estimated = -log(inv_z_opt_irl.^invbeta ./ (P' * inv_z_opt_irl)); 
    inv_l_hist(:,k_i) = inv_l_estimated;
    
    % state transition
    u_rand = rand(); 
    T_accum = P_accum(:, p_list(k_i));  
    for j_state = 1:m 
        if u_rand <= T_accum(j_state)
            p_list(k_i+1) = j_state;
            break;
        end
    end
    
    % update count
    a_counts(p_list(k_i+1)) = a_counts(p_list(k_i+1)) + 1;
    b_counts(p_list(k_i)) = b_counts(p_list(k_i)) + 1;
end

% plot Figure 10.4a
figure('Name','Figure10.4(a)');
stairs(0:k_bar_i, double(p_list) );
ylim([0.8, 4.2]);
xlim([0, 50]);
xticks([0, 50]);
xlabel('k');
yticks(1:4);
grid on;

% plot Figure 10.4b
figure('Name','Figure10.4(b)');
hold on;
for i_plot = 1:4 
    plot(inv_l_hist(i_plot,:)-inv_l_hist(1,:)+1, 'DisplayName', sprintf('$\\ell_{%d}$', i_plot));
end
ylim([0, 6]);
xlim([0, k_bar_i]);
xlabel('k');
legend('Interpreter','latex');
grid on;


% functions

function p_stationary = stationary_distribution(P)
    [V, D] = eig(P);
    [~, index] = max(diag(D));
    eigenvector_for_1 = V(:, index);
    sum_of_elements = sum(eigenvector_for_1);
    p_stationary = eigenvector_for_1 / sum_of_elements;
end

% Loss function for KL control, and it now optimizes V* directly
function sum_sq_residual = L_KL(V_star_var, beta_param, cost_val_param, P0_matrix) 
    term_log = log(P0_matrix' * exp(-beta_param * V_star_var));
    residuals = V_star_var - cost_val_param(:) + term_log;
    sum_sq_residual = sum(residuals.^2);
end

% Loss Function for IRL
function neg_log_likelihood = L_IRL(V_star_var, beta_param, a_counts_param, b_counts_param, P0_matrix) 
    term_log = log(P0_matrix' * exp(-beta_param * V_star_var));
    neg_log_likelihood = beta_param * a_counts_param' * V_star_var + b_counts_param' * term_log;
end

File: python/Figure10_2to4.py
# Author: Kenji Kashima
# Date  : 2025/09/10
# Note  : You should install scipy first.

import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from matplotlib.ticker import MultipleLocator
import config

np.random.seed(13)


# P(i,j) = Prob(current state = j → next state = i)
P = np.array([[1/3,    1/3,      0,      0],
                [0  ,    1/3,    1/3,      0],
                [0  ,    1/3,    1/3,    1/3],
                [2/3,      0,    1/3,    2/3]])
m,n = P.shape

def stationary_distribution(P: np.ndarray) -> np.ndarray:
    """
    stationary distribution vector of stochastic matrix P
    """
    eigvals, eigvecs = np.linalg.eig(P)  # right eigenvector
    index = np.argmax(np.real(eigvals))    # index for eig 1

    eigenvector_for_1 = np.real(eigvecs[:, index])
    # normalization
    p_stationary = eigenvector_for_1 / np.sum(eigenvector_for_1)
    return p_stationary

def figure10_2b(k_bar= 300):
    figsize = config.global_config(type= 1)
    # Figure 10.2 (a) transition probability
    
    p_stationary = stationary_distribution(P)
    print("p_stationary for P^0：", p_stationary)

    # p_stable = np.ones((4,1))/4
    # for _ in range(100):
    #     p_stable =  P@p_stable
    # print("p_100=",p_stable)
    
    # accumulated transition probability
    # P_accum(i,j) = Prob(current state = j → next state <= i)
    P_accum = np.cumsum(P, axis=0)

    state_list = np.zeros(k_bar, dtype=np.int64) 
    state_list[0] = 4 # start at 4

    # simulation of autonomous Markov Chain
    for k in range(1,k_bar):
        u = np.random.rand()                    # uniform distribution on [0,1]
        T = P_accum[:,int(state_list[k-1])-1]   # T(i) = Prob( next state <= i )
        for i in range(m):
            if u <= T[i]:
                state_list[k] = i+1
                break

    plt.figure(figsize=figsize)
    plt.stairs(state_list,linewidth=2,zorder = 10,baseline=4)
    plt.ylim([0.8,4.2])
    plt.xlim([0,k_bar])
    plt.xticks(np.arange(0,k_bar+1,10))
    plt.xlabel(r"$k$")
    plt.gca().yaxis.set_major_locator(MultipleLocator(1))
    plt.grid(axis='y')
    plt.tight_layout()
    plt.savefig("./Figure10_2b.pdf")
    plt.show()


def L_KL(v, beta, cost, P):
    """ L_KL(v) % find the solution of equation (10.20)"""
    return np.sum((v - cost + np.log(P.T @ np.exp(- beta*v)))**2)
         
def L_IRL(v, beta, a, b, P):
    """ L_IRL(v)  % find the solution of equation below (10.29)"""
    return beta * a @ v + b @ np.log(P.T @ np.exp(-beta * v))

def figure10_4(k_bar= 1000, sigma = 1.0):
    
    beta = 0.8
    invbeta = 1/beta
    cost = np.array([1, 2, 3, 4]) * sigma

    # find optimal transition probability P^pi
    results = minimize( fun = L_KL,
                        x0 = np.ones(4)*5,
                        args = (beta, cost, P),
                        method = "L-BFGS-B",
                        options = {
                            'gtol': 1e-10,    
                            'ftol': 1e-14,    
                            'maxiter': 1000, 
                            'maxls': 50       
                        }
                    )
    print("error:",results.fun)
    print("V*:",results.x)
    z_opt = np.exp(-beta*results.x) 
    P_opt = np.zeros((m,n))
    for i in range(m):
        for j in range(n):
            P_opt[i,j] = P[i,j] / (P[:,j]@(z_opt/z_opt[i]))
    
    print("sigma:",sigma, "Ppi:\n", P_opt)

    p_opt_stationary = stationary_distribution(P_opt)
    print("p_stationary for P_opt：", p_opt_stationary)

    P_accum = np.cumsum(P_opt, axis=0)

    state_list = np.zeros(k_bar+1,dtype=np.int8) 
    state_list[0] = 4   # start at 4
    a, b=np.array([0,0,0,1]), np.array([0,0,0,0])   # state visitation counts
    inv_l = np.zeros(n) # estimated state cost
    inv_l_hist =np.zeros((4,k_bar))   # history of estimated state costs

    # Solve L_IRL = 0 with V[1] = 'offset'
    offset = 0

    for k in range(k_bar):
        results = minimize(fun=L_IRL,x0= np.array([3,4,10,10]), args=(beta,a,b,P),
                        constraints= {'type': 'eq', 'fun': lambda x: x[0]-offset} )
        results_v = results.x   # estimated V*
        inv_z_opt = np.exp( -beta*results_v )   # estimated Z* by eq. (10.21)  
        inv_l = results_v + np.log(P.T @ inv_z_opt)   # estimated l eq. (10.25)
        inv_l_hist[:,k] = inv_l # store history of estimated state costs

        # simulation of optimally controlled Markov Chain
        u = np.random.rand()                
        T = P_accum[:,int(state_list[k])-1]     
        for i in range(m):
            if u <= T[i]:
                state_list[k+1] = i+1
                break
        # state visitation counts
        a[state_list[k+1]-1] +=1 
        b[state_list[k]-1] +=1

    figsize = config.global_config(type=1)
    plt.figure(figsize=figsize)
    plt.stairs(state_list,linewidth=2,zorder = 10,baseline=4)
    plt.ylim([0.8,4.2])
    plt.xlim([0,50])
    plt.xticks(np.arange(0,50+1,10))
    
    plt.xlabel(r"$k$")
    plt.gca().yaxis.set_major_locator(MultipleLocator(1))
    plt.grid(axis='y')
    plt.tight_layout()
    plt.savefig("./Figure10_4a.pdf")

    figsize = config.global_config(type=1)
    plt.figure(figsize=figsize)
    
    # state cost for state 1 is fixed to 1
    inv_l_hist = inv_l_hist - inv_l_hist[0,:] + 1
    for i in range(4):
        plt.plot(inv_l_hist[i,:],label=r"$\ell_{}$".format(i+1))
    plt.ylim([0,6.5])
    plt.yticks(np.arange(0,6,1))
    plt.xlim([0,k_bar])
    plt.xlabel(r"$k$")
    plt.legend()
    plt.grid(axis='y')
    plt.tight_layout()
    plt.savefig("./Figure10_4b.pdf")
    plt.show()

if __name__ == '__main__':
    figure10_2b(k_bar=50)
    figure10_4(k_bar=1000,sigma=1.0)
